llap-server/src/java/org/apache/hadoop/hive/llap/io/api/impl/LlapCacheMetadataSerializer.java

```diff
@@ -0,0 +1,202 @@
+      builder = LlapDaemonProtocolProtos.CacheEntry.newBuilder().setFileKey(encodedFileKey).setCacheTag(ctb)
+          .setFilePath(path);
+
+      lookupMap.put(fileKey, builder);
+    }
+    return builder;
+  }
+
+  private LlapDaemonProtocolProtos.CacheTag.Builder encodeCacheTag(CacheTag cacheTag) {
+    LlapDaemonProtocolProtos.CacheTag.Builder ctb =
+        LlapDaemonProtocolProtos.CacheTag.newBuilder().setTableName(cacheTag.getTableName());
+    if (cacheTag instanceof CacheTag.PartitionCacheTag) {
+      CacheTag.PartitionCacheTag partitionCacheTag = (CacheTag.PartitionCacheTag) cacheTag;
+      ctb.addAllPartitionDesc(Arrays.asList((partitionCacheTag.getEncodedPartitionDesc())));
+    }
+    return ctb;
+  }
+
+  public void loadData(LlapDaemonProtocolProtos.CacheEntryList data) {
+    for (LlapDaemonProtocolProtos.CacheEntry ce : data.getEntriesList()) {
+      try {
+        loadData(ce);
+      } catch (IOException ex) {
+        LOG.warn("Skip cache entry load to the cache.", ex);
+      }
+    }
+  }
+
+  private void loadData(LlapDaemonProtocolProtos.CacheEntry ce) throws IOException {
+    CacheTag cacheTag = decodeCacheTag(ce.getCacheTag());
+    DiskRangeList ranges = decodeRanges(ce.getRangesList());
+    Object fileKey = decodeFileKey(ce.getFileKey());
+    try (LlapOrcCacheLoader llr = new LlapOrcCacheLoader(new Path(ce.getFilePath()), fileKey, conf, cache,
+        metadataCache, cacheTag, tracePool)) {
+      llr.init();
+      llr.loadFileFooter();
+      llr.loadRanges(ranges);
+    }
+  }
+
+  private DiskRangeList decodeRanges(List<LlapDaemonProtocolProtos.CacheEntryRange> ranges) {
+    DiskRangeList.CreateHelper helper = new DiskRangeList.CreateHelper();
+    ranges.stream().sorted(Comparator.comparing(LlapDaemonProtocolProtos.CacheEntryRange::getStart))
+        .forEach(r -> helper.addOrMerge(r.getStart(), r.getEnd(), false, false));
+    return helper.get();
+  }
+
+  private CacheTag decodeCacheTag(LlapDaemonProtocolProtos.CacheTag ct) {
+    return ct.getPartitionDescCount() == 0 ? CacheTag.build(ct.getTableName()) : CacheTag
+        .build(ct.getTableName(), ct.getPartitionDescList());
+  }
+
+  @VisibleForTesting
+  ByteString encodeFileKey(Object fileKey) throws IOException {
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    DataOutputStream dos = new DataOutputStream(baos);
+    if (fileKey instanceof SyntheticFileId) {
+      SyntheticFileId fk = (SyntheticFileId) fileKey;
+      fk.write(dos);
+    } else {
+      dos.writeLong((Long) fileKey);
+    }
+    return ByteString.copyFrom(baos.toByteArray());
+  }
+
+  @VisibleForTesting
+  Object decodeFileKey(ByteString encodedFileKey) throws IOException {
+    byte[] bytes = encodedFileKey.toByteArray();
+    DataInput in = new DataInputStream(new ByteArrayInputStream(bytes));
+    Object fileKey;
+    if (bytes.length == Long.BYTES) {
+      fileKey = in.readLong();
+    } else {
+      SyntheticFileId fileId = new SyntheticFileId();
```

Comment:
I'm fine with not writing an indicator part for file ID type, but let's add some more comments then to explain in what cases is this Long or Synthetic

Reply:
Added some javadoc to be able to use the @see annotation pointing to the actual logic.